### PR TITLE
BAVL-266 changes to support (CSV) data extracts for MI reporting purposes

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
The changes in this PR provides the ability to download CSV files for the history of court video bookings by the date of making the booking or the date of the hearing.

Probation meetings will be done in a separate PR.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBookingHistory.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  val historyType: String,
+
+  val prisonCode: String,
+
+  val courtDescription: String?,
+
+  val courtCode: String?,
+
+  val probationTeamDescription: String?,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBookingHistory.kt:32]: Including probation.  Easy enough to remove if peeps think shouldn't include.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBookingHistory.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * This entity represents a view projection for v_video_booking_history, and contains the history of a booking.
+ *
+ * It is read-only (immutable).
+ */
+@Entity
+@Immutable
+@Table(name = "v_video_booking_history")
+data class VideoBookingHistory(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBookingHistory.kt:18]: Does the naming of this work?  Would `VideoBookingEvent` be better?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/VideoBookingsToCsvService.kt [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  @Transactional(readOnly = true)
+  fun courtBookingsByBookingDateToCsv(fromDate: LocalDate, toDate: LocalDate, csvOutputStream: OutputStream) {
+    writeCourtBookingsToCsv(videoBookingHistoryRepository.findByDateOfBookingBetween(fromDate, toDate), csvOutputStream)
+  }
+
+  private fun writeCourtBookingsToCsv(bookings: Stream<VideoBookingHistory>, csvOutputStream: OutputStream) {
+    // TODO need to pull out location descriptions ...
+    val courtBookings = bookings.filter(VideoBookingHistory::isCourtBooking)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/VideoBookingsToCsvService.kt:32]: Would could filter only court bookings in the actual repo query ...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/VideoBookingsToCsvService.kt [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  val postStartTime: LocalDateTime?,
+  val postEndTime: LocalDateTime?,
+  val mainLocationName: String,
+  val preLocationName: String?,
+  val postLocationName: String?,
+) {
+  constructor(vbh: VideoBookingHistory, locations: Map<String, List<Location>>) : this(
+    vbh.timestamp,
+    vbh.videoBookingId,
+    // Old BVLS does not have CANCEL, it has DELETE instead
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/VideoBookingsToCsvService.kt:86]: Do we want this?  It keeps it inline with the existing extract.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CsvDataExtractionService.kt [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  "mainEndTime",
+  "preStartTime",
+  "preEndTime",
+  "postStartTime",
+  "postEndTime",
+  "mainLocationName",
+  "preLocationName",
+  "postLocationName",
+)
+data class CourtBookingDto(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CsvDataExtractionService.kt:65]: Old BVLS has an event ID.  We don't.  Does this really matter?  We could use the appointment history id ...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CsvDataExtractionService.kt [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      .asSequence()
+
+    csvMapper
+      .writer(csvMapper.schemaFor(CourtBookingDto::class.java).withHeader())
+      .writeValues(csvOutputStream.bufferedWriter())
+      .use { writer -> courtBookings.forEach(writer::write) }
+  }
+}
+
+@JsonPropertyOrder(
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CsvDataExtractionService.kt:47]: This maintains the order of the attrs inline with the old service.  Otherwise they would be sorted alphabetically.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/resource/CsvDataExtractionController.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  @Operation(description = "Return details of court video link bookings by hearing date in CSV format. Restrict the response to events occurring within 'days' of start-date.")
+  @PreAuthorize("hasAnyRole('BOOK_A_VIDEO_LINK_ADMIN')")
+  fun downloadCourtBookingsByHearingDate(
+    @RequestParam(name = "start-date", required = true)
+    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
+    @Parameter(description = "The earliest hearing date for which to return event details.", required = true)
+    startDate: LocalDate,
+    @RequestParam(name = "days")
+    @Parameter(description = "Return details of events occurring within this number of days of start-date")
+    days: Long = 7,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/resource/CsvDataExtractionController.kt:35]: We don't restrict the number of days.  We could potentially ask for everything ...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CsvDataExtractionService.kt [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    writeCourtBookingsToCsv(videoBookingHistoryRepository.findByMainDateBetween(fromDate, toDate), csvOutputStream)
+  }
+
+  @Transactional(readOnly = true)
+  fun courtBookingsByBookingDateToCsv(fromDate: LocalDate, toDate: LocalDate, csvOutputStream: OutputStream) {
+    writeCourtBookingsToCsv(videoBookingHistoryRepository.findByDateOfBookingBetween(fromDate, toDate), csvOutputStream)
+  }
+
+  private fun writeCourtBookingsToCsv(bookings: Stream<VideoBookingHistory>, csvOutputStream: OutputStream) {
+    val locationsByPrisonCode = mutableMapOf<String, List<Location>>()
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/CsvDataExtractionService.kt:33]: This is the very limited caching for locations, and I mean limited.  It would pay to pull this out properly.  I'd rather do this independently in another PR.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/repository/VideoBookingHistoryRepository.kt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  fun findByDateOfBookingBetween(fromDate: LocalDate, toDate: LocalDate): Stream<VideoBookingHistory>
+
+  @Query(
+    value = """
+      FROM VideoBookingHistory vbh 
+      WHERE vbh.mainDate >= :fromDate and vbh.mainDate <= :toDate
+      ORDER BY vbh.mainDate, vbh.mainStartTime
+    """,
+  )
+  fun findByMainDateBetween(fromDate: LocalDate, toDate: LocalDate): Stream<VideoBookingHistory>
```
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/repository/VideoBookingHistoryRepository.kt:27]: I think these are both correct.  Tim feel fee to correct me on this ...

`mainDate` = court hearing/probation meeting date.
`dateOfBooking` = when the booking was made/created.

[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/repository/VideoBookingHistoryRepository.kt:27]: I see.

It lacks the second part of removing delete and create events.  That is not on the original JIRA, it probably should be ...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBookingHistory.kt (hunk 2 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBookingHistory.kt:19]: I think VideoBookingEvent works better.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/repository/VideoBookingHistoryRepository.kt [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  @QueryHints(
+    value = [
+      QueryHint(name = org.hibernate.jpa.HibernateHints.HINT_FETCH_SIZE, value = "" + Integer.MAX_VALUE),
+      QueryHint(name = org.hibernate.jpa.HibernateHints.HINT_CACHEABLE, value = "false"),
+      QueryHint(name = org.hibernate.jpa.HibernateHints.HINT_READ_ONLY, value = "true"),
+    ],
+  )
+  @Query(
+    value = """
+      FROM VideoBookingHistory vbh
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/repository/VideoBookingHistoryRepository.kt:39]: @smcveigh941 FYI

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBookingHistory.kt (hunk 2 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/entity/VideoBookingHistory.kt]: Done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/repository/VideoBookingHistoryRepository.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/repository/VideoBookingHistoryRepository.kt]: I think this is now sorted. Integration test has been updated.

**Review by timharrison-moj — APPROVED** [review]
LGTM
